<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

## Exercise 1 (Make me an odd list) ☻

In [13]:
for integer in range(10): #used for loope and to include range from 0 to 9
    if (integer**2) %2 == 0: #exclude the even products (0,2,4,16,36,64)
        continue
    print(integer **2)

1
9
25
49
81


In [10]:
list = [] #changed to create list
for integer in range(10):
    if (integer**2) %2 == 0:
        continue
    list.append(integer**2) #used append() to grow list
list

[1, 9, 25, 49, 81]

## Exercise 2 (Make me another odd list) ☻

In [19]:
[number**2 for number in range(10) if (number**2) %2 == 1] #only if it is odd numbers (remainder is 1)

[1, 9, 25, 49, 81]

## Exercise 3 (Time me!) ☻

In [1]:
%%timeit

list = [] #changed to create list
for integer in range(10):
    if (integer**2) %2 == 0:
        continue
    list.append(integer**2) #used append() to grow list
#list

1.08 µs ± 28.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [2]:
%%timeit

[number**2 for number in range(10) if (number**2) %2 == 1]

1.15 µs ± 66.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


loops append is faster (1.08 µs ± 28.5 ns per loop) compared to (1.15 µs ± 66.4 µs per loop)

## Exercise 4 (A problem of decay) ☻

In [ ]:
sample = 100
year = 0

while sample > 50: #for half the original amount
    sample * 0.95 #degrades by 5%, #without the =, it just did a normal multiplication and did not update the variable, loop continues indefinitely
    year += 1 # calculate each year
print(f" it takes {year} years to decay")

In [3]:
sample = 100
year = 0

while sample > 50: #for half the original amount
    sample *= 0.95 #degrades by 5%
    year += 1 # calculate each year
print(f" it takes {year} years to decay")

 it takes 14 years to decay


## Exercise 5 (Changes in CO) ☻

In [13]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

In [6]:
concentrations[i+1]

TypeError: 'float' object is not subscriptable

In [26]:
for year, concentrations in co2_data:
    for i in range(len(co2_data)):
        while concentrations[i+1] - concentrations[i] >= 3: #met this issue where each iteration gives the whole tuple
            print(year)

TypeError: 'float' object is not subscriptable

In [24]:
for i in range(1, len(co2_data)): #struggled to find out a way to use for year, concentrations as a way to unpack
    year, concentration = co2_data[i]  #chatgpt used this method to compare (which is not a good method as mentioned in loops_need)
    prev_year, prev_concentration = co2_data[i - 1] 
    if concentration - prev_concentration >= 3:
        print(f" the year {year} had a {concentration - prev_concentration:.2f} increase from the previous year") #added .2f to limit to 2dp

 the year 2016 had a 3.41 increase from the previous year
